### PR TITLE
Replace check_max function with resourceLimits directive

### LINKED ISSUE #2923: Replace `check_max` with `resourceLimits`
Once https://github.com/nextflow-io/nextflow/pull/2911 is in a stable release (24.04.0) we should update the template + all configs to use `resourceLimits` instead of our custom `check_max` function.

### PR DESCRIPTION
**Pull request by jfy133**
Closes #2923 

## PR checklist

- [ ] This comment contains a description of changes (with reason)
- [ ] `CHANGELOG.md` is updated
- [ ] If you've fixed a bug or added code that should be tested, add tests!
- [ ] Documentation in `docs` is updated

### CODE REVIEW

File: CHANGELOG.md
Comment (jfy133) [inline CHANGELOG.md:45]:
```suggestion
- Replaces the old custom `check_max()` function with the Nextflow native `resourceLimits` directive ([#3037](https://github.com/nf-core/tools/pull/3037))
```

**Comment by jfy133** [comment]
I tested locally by running the `nf-core/test-pipeline` example in the CI.

By adjusting `--max_memory` to 16, the FASTQC process that would normally get 32GB (label_medium) was indeed capped at 16 (normally it would be allocated 32)

image

**Comment by maxulysse** [comment]
> Let's just remove `params.max_*` entirely 🔥 

Just don't remove me please

**Comment by jfy133** [comment]
OK the tasks from the slack thread [here](https://nfcore.slack.com/archives/CE5LG7WMB/p1719320972375759)

- [ ] Remove all references to `--max_*` parameters
- [ ] Add lint tests (so people remove the parameters)
- [ ] Set new defaults in tests configs
- [ ] Update the docs on the website
  - [ ] Add warning about legacy useage
  - [ ] New docs to describe how to use new resourceDirectives 
- [ ] mass update all configs on nf-core/configs
    - [ ] Keep old `--max_*` parameters (buit leave comment)
    - [ ] Copy vbalues to new resourcesLimits (with

**Comment by jfy133** [comment]
@nf-core-bot fix linting

File: nf_core/pipeline-template/nextflow_schema.json
Comment (jfy133) [inline nf_core/pipeline-template/nextflow_schema.json:137]:
```suggestion
        },
        {%- endif %}
```

File: tests/pipelines/lint/test_nextflow_config.py
Comment (mirpedrol) [inline tests/pipelines/lint/test_nextflow_config.py:72]: Instead of removing this test we should change it to use a different variable. `max_multiqc_email_size`, `publish_dir_mode` or `custom_config_version` would be good candidates

File: tests/pipelines/lint/test_nextflow_config.py
Comment (mirpedrol) [inline tests/pipelines/lint/test_nextflow_config.py:101]: Same as above, we should use a different parameter but keep the test

File: tests/pipelines/lint/test_nextflow_config.py
Comment (mirpedrol) [inline tests/pipelines/lint/test_nextflow_config.py:116]: same as above, keep test

File: tests/pipelines/lint/test_nextflow_config.py
Comment (mirpedrol) [inline tests/pipelines/lint/test_nextflow_config.py:127]: same as above, keep test

**Comment by mirpedrol** [comment]
This is ready to merge, ok if we merge it @jfy133 ?
We have to change the branch protection for the two pending tests, but I want to do it right before merging, otherwise it will block other PRs.

**Comment by mirpedrol** [comment]
ok I will merge it now! It won't affect old pipelines so it should be good 👍

**Review by mirpedrol — APPROVED** [review]

**Pull request merged**
